Public/Src/Pips/Dll/Operations/Process.cs

```diff
@@ -331,6 +342,18 @@ public sealed partial class Process : Pip
         [PipCaching(FingerprintingRole = FingerprintingRole.Semantic)]
         public ReadOnlyArray<AbsolutePath> PreserveOutputWhitelist { get; }
 
+        /// <summary>
+        /// Source Change Affected Input Files
+        /// </summary>
+        [PipCaching(FingerprintingRole = FingerprintingRole.None)]
+        public ReadOnlyArray<AbsolutePath> SourceChangeAffectedOutputFiles { get; set; }
+
+        /// <summary>
+        /// Source Change Affected Input Directories
+        /// </summary>
+        [PipCaching(FingerprintingRole = FingerprintingRole.None)]
+        public ReadOnlyArray<AbsolutePath> SourceChangeAffectedOutputDirectories { get; set; }

```

Comment:
Why do we need these two? Also we don't allow you to modify the Process once it is created. So, the property `set` method should not be there.

Reply:
Yes, the Process object should indeed be immutable, it can also be paged out during builds or analysis and then these fields aren't saved, only the fields added during graph construction are saved to the byte array and then the pips are hydrated as needed (potentially many times)

---
In reply to: [314804613](https://github.com/microsoft/BuildXL/pull/748#discussion_r314804613) [](ancestors = 314804613)

Reply:
This two are set at the time when master distribute the pip. Master will retrieve the global impacted outputs with the current state of the build. But the process is created earlier than the distribution, that's why I need the set here.

Reply:
I changed the implementation. I compute the change affected input in the master and pass the result in the processRunnable instead of process.